README
Closure
=======

Closure is a shutdown utility for a composite X windowing environment.

Project site: http://code.google.com/p/closure/


Dependencies
============
These are the versions I have installed, although Closure will most likey work
with versions less then these. I'll try to update this list with the actual
versions shortly.
	>= pygtk-2.10 (most likely works with 2.8, haven't tested)
	>= pycairo-1.2
	>= librsvg-2.16
	>= gnome-python-2.16
	>= gnome-python-desktop-2.16 (with rsvg support)
	>= libglade-2.6
	>= gconf-2.14

	Second level dependencies
	-------------------------
	>= glib-2.12
	>= gtk+-2.10  (most likely works with 2.8, haven't tested)
	>= cairo-1.2 (with SVG support, and optional GLITZ support)
	>= pango-1.14
	>= libgnome-2.16
	>= gnome-desktop-2.16


License
=======
  Copyright (C) 2007 Anthony Arobone <aarobone@gmail.com>

  This program is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.

Installation
============
./autogen.sh
[./configure --help] (to get/set optional configure settings)
make
make install (as root)

	Installation Notes:
	------------------
	1) Even though the configure script prints out the following:

	   Using $(sysconfdir)/gconf/schemas as install directory for schema files

	   The path used is always /etc/gconf/schemas. I'm not sure how to change 
	   this printout message yet ;)


Troubleshooting
===============

1) If you are getting any GConf errors when attempting to run Closure, please
double check that the gconf closure schema was installed successfully. You
can use gcond-editor to check this. If the schema needs to be installed
execute the following command from the root closure dir:
	gconftool-2 --install-schema-file=src/closure.schemas




NEWS
Closure project


src/SVGButton.py
#!/usr/bin/env python
#
#  Copyright (C) 2007 Anthony Arobone <aarobone@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.
#
#  Author: Anthony Arobone <aarobone@gmail.com>
#
#  Notes: Closure SVGButton widget
#

import gobject
import pygtk
pygtk.require('2.0')
import gtk
import gtk.gdk
import cairo
import rsvg
import os
import math
import pango
import pangocairo

#
# Class to handle doing SVG Image Buttons
#
class SVGButton(gtk.EventBox):
	"SVGButton Widget"

	width = 48
	height = 48
	text_width = 0
	text_height = 0
	text_offset = 2
	
	def __init__( self, scale=1, text=None, bin=None ):
		gtk.EventBox.__init__(self)


		# instance variables
		self.text_layout = None
		self.bin = None
		if bin:
			self.bin = bin.split( " " )
			
		self.set_size( 48, 48 )
		self.scale = scale
		self.scale_background = self.scale + 1 # normal scale
		if text != None and len(text) > 0:
			self.scale_background = self.scale + 1.5 # make background big enough to fit text
			self.text_layout = self.create_pango_layout( text )
			self.text_layout.set_font_description( pango.FontDescription( "Sans 16" ) )
			self.text_width, self.text_height = self.text_layout.get_pixel_size()
			self.set_size( self.text_width, self.text_height )

		self.inside = False
		self.svg_background = None
		self.svg_foreground = None
		self.svg_focus = None
		self.pressed = False


		# tell widget we want to draw it ourself
		self.set_app_paintable( True )
		self.set_visible_window( False )

		# turn on these events
		self.add_events( gtk.gdk.BUTTON_PRESS_MASK | gtk.gdk.BUTTON_RELEASE_MASK | \
						gtk.gdk.ENTER_NOTIFY_MASK | gtk.gdk.LEAVE_NOTIFY_MASK | \
						gtk.gdk.KEY_PRESS_MASK | gtk.gdk.KEY_RELEASE_MASK | \
						gtk.gdk.FOCUS_CHANGE_MASK )

		# enable focus 
		self.set_flags( gtk.CAN_FOCUS )
		
		# register callbacks for events
		self.connect( "screen-changed", self.screen_changed )
		self.connect( "configure-event", self.do_configure )
		self.connect( "expose-event", self.do_expose )
		self.connect( "button-press-event", self.press )
		self.connect( "button-release-event", self.release )
		self.connect( "enter-notify-event", self.do_enter )
		self.connect( "leave-notify-event", self.do_leave )
		


	def screen_changed( self, widget, old_screen=None ):
#		print "SVGButton screen changed"
		screen = widget.get_screen()
		colormap = screen.get_rgba_colormap()
		widget.set_colormap( colormap )

		if self.text_layout:
			self.text_layout.context_changed()

		return False


	#
	# Callback to manage widget changes, like resize
	#
	def do_configure( self, widget, event ):
		print "SVGButton configure"
		pass


	#
	# Callback to draw this widget
	#
	def do_expose( self, widget, event ):
		#print "SVGButton expose"

		rect = self.get_allocation()
		x = rect.x
		y = rect.y
		w = rect.width
		h = rect.height
#		print w, "x", h
		(w, h) = self.get_size()

		# get cairo context
		cr = self.window.cairo_create()
		
		# Debugging lines
#		cr.set_source_rgba( 1, 0, 0, 1 )
#		sw = self.get_screen().get_width()
#		sh = self.get_screen().get_height()
#		cr.move_to( sw/2, 0 )
#		cr.line_to( sw/2, sh/2 )
#		cr.line_to( sw, sh/2 )
#		cr.line_to( 0, sh/2 )
#		cr.stroke()
		
		# set clip region
#		print "expose area = ", event.area.x, ",", event.area.y, ",", event.area.width, ",", event.area.height
		cr.rectangle( event.area.x, event.area.y, event.area.width, event.area.height )
		cr.clip()
	
		# draw widget over the source
		cr.set_operator( cairo.OPERATOR_OVER )
		
		# calculate the base translation for this widget within it's parent drawing area
		if rect.height > rect.width:
			# if the physical height is bigger then the width (not square)
			# then offset all the drawing in the middle vertically
			diff = rect.width - rect.height
			diff = diff / 2
			cr.translate( x, y+diff )
		else:
			cr.translate( x, y )
		base_mat = cr.get_matrix()

#		print "button = ", w, "x", h, " scaled by ", self.scale

		# render background
		if self.inside and self.svg_background:
			cr.scale( self.scale_background, self.scale_background )
			self.svg_background.render_cairo( cr )
			cr.set_matrix( base_mat )
	
		# render foreground
		if self.svg_foreground:
			iw = self.svg_foreground.props.width * self.scale
			ih = self.svg_foreground.props.height * self.scale
			# center foreground
			cr.translate( (w-iw)/2, (h-ih)/2 )
			if self.text_layout:
				cr.translate( 0, self.text_height/-2 )
			cr.scale( self.scale, self.scale )
			self.svg_foreground.render_cairo( cr )
			cr.set_matrix( base_mat )

		# render focus layer
		if self.inside and self.svg_focus:
			iw = self.svg_focus.props.width * self.scale
			ih = self.svg_focus.props.height * self.scale
			# center focus
			cr.translate( (w-iw)/2, (h-ih)/2 )
			if self.text_layout:
				cr.translate( 0, self.text_height/-2 )
			cr.scale( self.scale, self.scale )
			self.svg_focus.render_cairo( cr )
			cr.set_matrix( base_mat )

		# render text
		if self.text_layout:
			# calc where the text should go
			fw, fh = self.text_layout.get_pixel_size()
			if self.svg_foreground:
				ih = self.svg_foreground.props.height * self.scale
				cr.translate( (w-fw)/2, (h-ih)/2 + ih)
			else:
				cr.translate( (w-fw)/2, h/2 )
			cr.translate( 0, self.text_height / -2 + self.text_offset )
			text_mat = cr.get_matrix()
			
			if True or self.inside:
				# black shadow
				cr.translate( 1, 2 ) # bottom right
				cr.set_source_rgba( 0.1, 0.1, 0.1, 1.0 )
				cr.update_layout( self.text_layout )
				cr.show_layout( self.text_layout )
				cr.set_source_rgba( 0.6, 0.6, 0.6, 1.0 )
				cr.translate( -2, -1 ) # bottom left
				cr.show_layout( self.text_layout )
				cr.translate( 0, -1 ) # left
				cr.show_layout( self.text_layout )
				cr.translate( 2, 0 ) # right
				cr.show_layout( self.text_layout )
				cr.translate( -1, -1 ) # top
				cr.show_layout( self.text_layout )

			# white 
			cr.set_matrix( text_mat )
			cr.set_source_rgba( 1.0, 1.0, 1.0, 1.0 )
			cr.update_layout( self.text_layout )
			cr.show_layout( self.text_layout )

		# don't let anyone else draw this widget after this method ends
		return True


	def do_enter( self, widget, event ):
#		print "SVGButton enter", event.mode
		if event.mode == gtk.gdk.CROSSING_NORMAL:
			self.inside = True
			rect = self.get_allocation()
			self.window.invalidate_rect( rect, False )
			self.window.process_updates( False )
		return True
		
		
	def do_leave( self, widget, event ):
#		print "SVGButton leave", event.mode
		if event.mode == gtk.gdk.CROSSING_NORMAL:
			self.inside = False
			rect = self.get_allocation()
			self.window.invalidate_rect( rect, False )
			self.window.process_updates( False )
		return True
		
	
	def press(self, widget, event ):
#		print "SVGButton press "
		if event.button == 1:
			self.pressed = True

#			rect = self.get_allocation()
#			self.window.invalidate_rect( rect, False )
#			self.window.process_updates( False )
		return True


	def release(self, widget, event ):
		retval = True
		if event.button == 1:
			if self.inside:
				# do button action here
#				print "SVGButton release"
				
				if self.bin:
					# if the binary is empty string, just exit the program
					if self.bin[0] == "":
						gtk.main_quit()
						return retval
					
					# replace current process [Clean up needed? Got me]
					print self.bin, len(self.bin)
					if len(self.bin) > 1:
						os.execvp( self.bin[0], self.bin )
					else:
						os.execlp( self.bin[0], self.bin[0] )
					
					# stop closure, and start process
#					gtk.main_quit()
#					print os.spawnlp(os.P_NOWAIT, self.bin, " " )
				
					retval = False

		self.pressed = False

#		rect = self.get_allocation()
#		self.window.invalidate_rect( rect, False )
#		self.window.process_updates( False )
		return retval


	def set_background_svg( self, svg ):
		self.svg_background = svg
		
		# on set background, resize the widget
		self.set_size( svg.props.width * self.scale_background, svg.props.height * self.scale_background)


	def set_foreground_svg( self, svg ):
		self.svg_foreground = svg

		# on set foreground, resize the widget
		# don't forget to add on the height of the text
		self.set_size( svg.props.width * self.scale, svg.props.height * self.scale + self.text_height )
			
	
	def set_focus_svg( self, svg ):
		self.svg_focus = svg

		# on set focus, resize the widget
		# don't forget to add on the height of the text
		self.set_size( svg.props.width * self.scale, svg.props.height * self.scale + self.text_height )
		

	#
	# Sets the size of this button
	# Only grows, never shrinks
	#
	def set_size( self, w, h ):
		# account for floats
		self.width = max( self.width, int(math.ceil( w )) )
		self.height = max( self.height, int( math.ceil( h )) )
		self.set_size_request( self.width, self.height )

	#
	# Gets the size of this button
	#
	def get_size( self ):
		return ( self.width, self.height )


src/common.in.py
#!/usr/bin/env python
#
#  Copyright (C) 2007 Neil Jagdish Patel <njpatel@gmail.com>
#					  Anthony Arobone <aarobone@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.
#
#  Author: Neil Jagdish Patel <njpatel@gmail.com>
#  Contributor: Anthony Arobone <aarobone@gmail.com>
#
#  Notes: Closure commons
#         Taken from avant-window-navigator's preferences
#

try:
 	import pygtk
  	pygtk.require("2.0")
except:
  	pass
try:
	import gtk
except:
	sys.exit(1)


DATA_DIR = "@PKGDATADIR@"


# GCONF KEYS

GENERAL_PATH			= "/apps/closure/general"
GENERAL_FULLSCREEN		= "/apps/closure/general/fullscreen"		# bool

BACKGROUND_PATH			= "/apps/closure/background"
BACKGROUND_STYLE		= "/apps/closure/background/style" #bool
BACKGROUND_SOLID		= "/apps/closure/background/solid" #string
BACKGROUND_LINEAR_STEP_1	= "/apps/closure/background/linear_step_1" #string
BACKGROUND_LINEAR_STEP_2	= "/apps/closure/background/linear_step_2" #string
BACKGROUND_LINEAR_STEP_3	= "/apps/closure/background/linear_step_3" #string
BACKGROUND_SVG_IMAGE		= "/apps/closure/background/svg" 		#string

BUTTONS_PATH		= "/apps/closure/buttons"
BUTTONS_BG_SVG		= "/apps/closure/buttons/bg_svg"				#string
BUTTONS_CANCEL_ENABLED	= "/apps/closure/buttons/cancel_enabled"	#bool
BUTTONS_CANCEL_LABEL	= "/apps/closure/buttons/cancel_label"		#string
BUTTONS_CANCEL_SVG		= "/apps/closure/buttons/cancel_svg" 		#color
BUTTONS_LOCK_ENABLED	= "/apps/closure/buttons/lock_enabled"		#bool
BUTTONS_LOCK_LABEL		= "/apps/closure/buttons/lock_label"		#string
BUTTONS_LOCK_SVG		= "/apps/closure/buttons/lock_svg" 			#color
BUTTONS_LOGOUT_ENABLED	= "/apps/closure/buttons/logout_enabled"	#bool
BUTTONS_LOGOUT_LABEL	= "/apps/closure/buttons/logout_label"		#string
BUTTONS_LOGOUT_SVG		= "/apps/closure/buttons/logout_svg" 		#color
BUTTONS_REBOOT_ENABLED	= "/apps/closure/buttons/reboot_enabled"	#bool
BUTTONS_REBOOT_LABEL	= "/apps/closure/buttons/reboot_label"		#string
BUTTONS_REBOOT_SVG		= "/apps/closure/buttons/reboot_svg" 		#color
BUTTONS_SHUTDOWN_ENABLED	= "/apps/closure/buttons/shutdown_enabled"	#bool
BUTTONS_SHUTDOWN_LABEL		= "/apps/closure/buttons/shutdown_label"	#string
BUTTONS_SHUTDOWN_SVG		= "/apps/closure/buttons/shutdown_svg" 		#color
BUTTONS_SCALE				= "/apps/closure/buttons/scale"				#float

COMMANDS_PATH		= "/apps/closure/commands"
COMMANDS_LOCK		= "/apps/closure/commands/lock"		#string
COMMANDS_LOGOUT		= "/apps/closure/commands/logout"	#string
COMMANDS_REBOOT		= "/apps/closure/commands/reboot"	#string
COMMANDS_SHUTDOWN	= "/apps/closure/commands/shutdown"	#string


COLOR_BIT_MAX = 65535




#
# Load a boolean from gconf
#
def loadBool(client, key, default):
	v = client.get_bool( key )
	if v == None:
		return default
	else:
		return v

#
# Load a float from gconf
#
def loadFloat(client, key, default):
	v = client.get_float( key )
	if v == None:
		return default
	else:
		return v


#
# Load a string from gconf
#
def loadString(client, key, default):
	v = client.get_string( key )
	if v == None:
		return default
	else:
		return v
	
#
# Load a string from gconf
#
# Returns: A 4-tuple containing floats for RGBA color.
#
def loadColor(client, key, default):
	v = client.get_string( key )
	if v == None:
		v = default
	(color, alpha) = make_color( v )
	r = color.red / float(COLOR_BIT_MAX)
	g = color.green / float(COLOR_BIT_MAX)
	b = color.blue / float(COLOR_BIT_MAX)
	a = alpha / float(COLOR_BIT_MAX)

	return (r, g, b, a)




def dec2hex(n):
	"""return the hexadecimal string representation of integer n"""
	n = int(n)
	if n == 0:
		return "00"
	return "%0.2X" % n
 
 
def hex2dec(s):
	"""return the integer value of a hexadecimal string s"""
	return int(s, 16)


def make_color(hexi):
	"""returns a gtk.gdk.Color from a hex string RRGGBBAA"""
	color = gtk.gdk.color_parse('#' + hexi[:6])
	alpha = hex2dec(hexi[6:])
	alpha = (float(alpha)/255)* COLOR_BIT_MAX
	return color, int(alpha)


def make_color_string(color, alpha):
	"""makes readable string from gdk.color & alpha (0-65535) """
	string = ""
	
	string = string + dec2hex(int( (float(color.red) / COLOR_BIT_MAX)*255))
	string = string + dec2hex(int( (float(color.green) / COLOR_BIT_MAX)*255))
	string = string + dec2hex(int( (float(color.blue) / COLOR_BIT_MAX)*255))
	string = string + dec2hex(int( (float(alpha) / COLOR_BIT_MAX)*255))
	
	#hack
	return string	


src/closure.py
#!/usr/bin/env python
#
#  Copyright (C) 2007 Anthony Arobone <aarobone@gmail.com>
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301  USA.
#
#  Author: Anthony Arobone <aarobone@gmail.com>
#
#  Notes: Closure app
#


import gobject
import pygtk
pygtk.require('2.0')
import gtk
import cairo
import rsvg
import gconf
import os
import SVGButton
import common
import closure_preferences


class Closure:
	"Closure"
	
	width = 640
	height = 480
	# fullscreen mode
	#	NOTE: fullscreen mode causes slowness with mouse presses, not sure why.
	do_fullscreen = False
	background_style = None
	background_solid = None
	background_linear_1 = None
	background_linear_2 = None
	background_linear_3 = None
	background_svg = None
	background_svg_scale = 1
	button_background = None
	lock_svg = None
	logout_svg = None
	reboot_svg = None
	shutdown_svg = None
	cancel_svg = None
	cancel_focus_svg = None


	def __init__(self):
		self.window = gtk.Window(gtk.WINDOW_TOPLEVEL)
		self.window.set_title( "Closure" )
		try:
			self.window.set_icon_from_file( os.path.join( common.DATA_DIR, "closure.svg") )
		except:
			print "Icon closure.svg not found."

		self.width = self.window.get_screen().get_width()
		self.height = self.window.get_screen().get_height()
		self.window.set_position( gtk.WIN_POS_CENTER )
		self.window.set_default_size( self.width, self.height ) 
		self.window.set_geometry_hints( None, self.width, self.height )

		self.client = gconf.client_get_default()
		self.client.add_dir( common.GENERAL_PATH, gconf.CLIENT_PRELOAD_NONE )
		self.client.add_dir( common.BACKGROUND_PATH, gconf.CLIENT_PRELOAD_NONE )
		self.client.add_dir( common.BUTTONS_PATH, gconf.CLIENT_PRELOAD_NONE )
		self.client.add_dir( common.COMMANDS_PATH, gconf.CLIENT_PRELOAD_NONE )

	
		# set fullscreen mode
		self.do_fullscreen = self.client.get_bool( common.GENERAL_FULLSCREEN )
		if self.do_fullscreen:
			self.window.fullscreen()

		# register callbacks
		self.window.connect( "expose-event", self.expose )
		self.window.connect( "screen-changed", self.screen_changed )
		self.window.connect( "key-press-event", self.keypressed)
		self.window.connect( "destroy", self.destroy)
		
	
		# set alpha
		self.window.set_app_paintable( True )
		self.screen_changed( self.window )
		
		self.window.set_resizable( False )
		self.window.set_keep_above( True )
		#self.window.set_urgency_hint( True )
		self.window.set_decorated( False )
		self.window.set_gravity( gtk.gdk.GRAVITY_CENTER )
	
		# load background properties
		self.background_style = common.loadString( self.client, common.BACKGROUND_STYLE, "linear" )
		self.background_solid = common.loadColor( self.client, common.BACKGROUND_SOLID, "000000AA" )
		self.background_linear_1 = common.loadColor( self.client, common.BACKGROUND_LINEAR_STEP_1, "000000E6" )
		self.background_linear_2 = common.loadColor( self.client, common.BACKGROUND_LINEAR_STEP_2, "000000BF" )
		self.background_linear_3 = common.loadColor( self.client, common.BACKGROUND_LINEAR_STEP_3, "00000080" )

		# load SVG images
		self.loadSVGImages()
	

		# Read button properties
		button_scale = common.loadFloat( self.client, common.BUTTONS_SCALE, 1 )
		show_lock = common.loadBool(  self.client, common.BUTTONS_LOCK_ENABLED, True )
		lock_label = common.loadString(  self.client, common.BUTTONS_LOCK_LABEL, "Lock" )
		lock_cmd = common.loadString(  self.client, common.COMMANDS_LOCK, None )
		
		show_logout = common.loadBool(  self.client, common.BUTTONS_LOGOUT_ENABLED, False )
		logout_label = common.loadString(  self.client, common.BUTTONS_LOGOUT_LABEL, "Logout" )
		logout_cmd = common.loadString(  self.client, common.COMMANDS_LOGOUT, None )

		show_reboot = common.loadBool(  self.client, common.BUTTONS_REBOOT_ENABLED, True )
		reboot_label = common.loadString(  self.client, common.BUTTONS_REBOOT_LABEL, "Reboot" )
		reboot_cmd = common.loadString(  self.client, common.COMMANDS_REBOOT, None )

		show_shutdown = common.loadBool(  self.client, common.BUTTONS_SHUTDOWN_ENABLED, True )
		shutdown_label = common.loadString(  self.client, common.BUTTONS_SHUTDOWN_LABEL, "Shutdown" )
		shutdown_cmd = common.loadString(  self.client, common.COMMANDS_SHUTDOWN, None )

		show_cancel = common.loadBool(  self.client, common.BUTTONS_CANCEL_ENABLED, True )
		cancel_label = common.loadString(  self.client, common.BUTTONS_CANCEL_LABEL, None )
	
		# create SVG Buttons
		if show_lock:
			self.button_svg_lock = SVGButton.SVGButton( button_scale, lock_label, lock_cmd )
			self.button_svg_lock.set_foreground_svg( self.lock_svg )
			self.button_svg_lock.set_background_svg( self.button_background )
			self.button_svg_lock.set_flags( gtk.CAN_DEFAULT )

		if show_logout:
			self.button_svg_logout = SVGButton.SVGButton( button_scale, logout_label, logout_cmd )
			self.button_svg_logout.set_foreground_svg( self.logout_svg )
			self.button_svg_logout.set_background_svg( self.button_background )

		if show_reboot:
			self.button_svg_reboot = SVGButton.SVGButton( button_scale, reboot_label, reboot_cmd )
			self.button_svg_reboot.set_foreground_svg( self.reboot_svg )
			self.button_svg_reboot.set_background_svg( self.button_background )

		if show_shutdown:
			self.button_svg_halt = SVGButton.SVGButton( button_scale, shutdown_label, shutdown_cmd )
			self.button_svg_halt.set_foreground_svg( self.shutdown_svg )
			self.button_svg_halt.set_background_svg( self.button_background )

		if show_cancel:
			self.button_cancel = SVGButton.SVGButton( 1, cancel_label, " " )
			self.button_cancel.set_foreground_svg( self.cancel_svg )
			self.button_cancel.set_focus_svg( self.cancel_focus_svg )


		# add buttons to HBox
		self.box = gtk.HButtonBox()
		self.box.set_layout( gtk.BUTTONBOX_SPREAD )
		self.box.set_spacing( 0 )
		expand = False
		fill = False
		self.box.pack_start(self.button_svg_lock, expand, fill, 0)
		if show_logout:
			self.box.pack_start(self.button_svg_logout, expand, fill, 0)
		self.box.pack_start(self.button_svg_reboot, expand, fill, 0)
		self.box.pack_start(self.button_svg_halt, expand, fill, 0)


		# add HBox to alignment and to main window
		self.align = gtk.Alignment( 0.5, 0.5, 0.3, 0 )
		self.align.add( self.box )
		
		con = gtk.Table( 2 )
		con.attach( self.align, 1, 2, 1, 2, gtk.EXPAND | gtk.FILL, gtk.EXPAND )
		if show_cancel:
			con.attach( self.button_cancel, 1, 2, 2, 3, gtk.SHRINK, gtk.SHRINK )
		self.window.add( con )

	
		self.window.show_all()



	#
	# Loads all SVGs closure might need
	#
	def loadSVGImages(self):
		svg = self.client.get_string( common.BACKGROUND_SVG_IMAGE )
		if svg:
			try:
				self.background_svg = rsvg.Handle( svg )
				
				# for now we'll uniformly scale the bg image to fit the screen
				# TODO: make this a closure-preferences setting, uniform and non-uniform
				dw = abs(self.width - self.background_svg.props.width)
				dh = abs(self.height - self.background_svg.props.height)
				if dw > dh:
					# the width is smaller then the height
					self.background_svg_scale = self.width / float(self.background_svg.props.width)
				else:
					# the height is smaller then the width
					self.background_svg_scale = self.height / float(self.background_svg.props.height)
			except:
				print "SVG file:", svg

		svg = self.client.get_string( common.BUTTONS_BG_SVG )
		if svg:
			try:
				self.button_background = rsvg.Handle( svg )
			except:
				print "SVG file:", svg
		
		svg = self.client.get_string( common.BUTTONS_LOCK_SVG )
		if svg:
			try:
				self.lock_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg

		svg = self.client.get_string( common.BUTTONS_LOGOUT_SVG )
		if svg:
			try:
				self.logout_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg

		svg = self.client.get_string( common.BUTTONS_REBOOT_SVG )
		if svg:
			try:
				self.reboot_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg

		svg = self.client.get_string( common.BUTTONS_SHUTDOWN_SVG )
		if svg:
			try:
				self.shutdown_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg
		
		svg = self.client.get_string( common.BUTTONS_CANCEL_SVG )
		if svg:
			try:
				self.cancel_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg
			svg = svg.replace(".svg", "-focus.svg")
			try:
				self.cancel_focus_svg = rsvg.Handle( svg )
			except:
				print "SVG file:", svg
	

	def expose( self, widget, event):
		cr = widget.window.cairo_create()

		# set clip region
		#print "expose area = ", event.area.x, ",", event.area.y, ",", event.area.width, ",", event.area.height
		cr.rectangle( event.area.x, event.area.y, event.area.width, event.area.height )
		cr.clip()

		(w, h) = widget.window.get_size()
		cr.set_operator( cairo.OPERATOR_SOURCE )

		# radial gradient attempt, not so hot
#		pat = cairo.RadialGradient( w/2, h/2, 10, w/2, h/2, w )
#		pat.add_color_stop_rgba( 0.0, 1.0, 0.0, 0.0, 0.8 )
#		pat.add_color_stop_rgba( 1.0, 1.0, 1.0, 1.0, 0.3 )
#		cr.set_source( pat )
#		cr.paint()
#
#		cr.set_operator( cairo.OPERATOR_OVER )
#		return False

		# SVG image background, better make something worthwhile
		if self.background_style == "svg":
			if self.background_svg:
				if self.background_svg_scale != 1:
					cr.scale( self.background_svg_scale, self.background_svg_scale )
				self.background_svg.render_cairo( cr )
			else:
				# if svg does not exist fall back to solid bg
				self.background_style = "solid"

		# solid black, decent
		if self.background_style == "solid":
#			cr.set_source_rgba( 0.1, 0.1, 0.1, 0.5 )
			cr.set_source_rgba( self.background_solid[0], self.background_solid[1], self.background_solid[2], self.background_solid[3] )
			cr.paint()

		# diagonal linear gradient, ok
		if self.background_style == "linear":
			pat = cairo.LinearGradient( 0.0, 0.0, w/1.5, h )
#			pat.add_color_stop_rgba( 0.1, 0.0, 0.0, 0.0, 0.9 )
#			pat.add_color_stop_rgba( 0.6, 0.0, 0.0, 0.0, 0.75 )
#			pat.add_color_stop_rgba( 1.0, 0.0, 0.0, 0.0, 0.5 )
			pat.add_color_stop_rgba( 0.1, self.background_linear_1[0], self.background_linear_1[1], self.background_linear_1[2], self.background_linear_1[3] )
			pat.add_color_stop_rgba( 0.6, self.background_linear_2[0], self.background_linear_2[1], self.background_linear_2[2], self.background_linear_2[3] )
			pat.add_color_stop_rgba( 1.0, self.background_linear_3[0], self.background_linear_3[1], self.background_linear_3[2], self.background_linear_3[3] )
			cr.set_source( pat )
			cr.paint()
		
		# radial gradient attempt, not so hot
#		pat = cairo.RadialGradient( 200, 200, 10, 200, 200, w )
#		pat.add_color_stop_rgba( 0.0, 0.0, 0.0, 0.0, 0.8 )
#		pat.add_color_stop_rgba( 1.0, 0.0, 0.0, 0.0, 0.3 )
#		cr.set_source( pat )
#		cr.paint()

		return False

	def screen_changed( self, widget, old_screen=None):
		screen = widget.get_screen()
		colormap = screen.get_rgba_colormap()
		widget.set_colormap( colormap )
		return False


	def keypressed(self, widget, event, data=None):
		if event.keyval == 96: #tilda
			self.window.destroy()
			app = closure_preferences.preferences()
			gtk.main()
		if event.keyval == 65307: #ESC
			gtk.main_quit()


	def destroy(self, widget, data=None):
		gtk.main_quit()


	def main(self):
		gtk.main()


if __name__ == "__main__":
	closure = Closure()
	closure.main()
